File: config/redisClient.js
// config/redisClient.js
const redis = require("redis");

let redisClient = null;

/**
 * Determines the Redis connection URL.
 * Uses REDIS_URL from environment if available.
 * Defaults to 'redis://redis:6379' if NODE_ENV is 'production' (Docker context).
 * Defaults to 'redis://localhost:6379' otherwise (local development context).
 * @returns {string} Redis connection URL
 */
const getRedisUrl = () => {
  if (process.env.REDIS_URL) {
    return process.env.REDIS_URL;
  }
  return process.env.NODE_ENV === "production"
    ? "redis://redis:6379"
    : "redis://localhost:6379";
};

/**
 * Initialize Redis client (singleton)
 * @returns {Promise<object>} Redis client
 */
const initRedisClient = async () => {
  if (redisClient) {
    return redisClient;
  }

  const redisUrl = getRedisUrl();
  console.log(`Attempting to connect to Redis at: ${redisUrl}`);

  redisClient = redis.createClient({
    url: redisUrl,
    socket: {
      keepAlive: 5000, // Send TCP keepalive packets every 5 seconds
      connectTimeout: 10000, // Abort connection attempt after 10 seconds
    },
    // Performance optimizations
    disableOfflineQueue: true, // Don't queue commands when offline
  });

  redisClient.on("error", (err) => console.error("Redis Client Error:", err));
  redisClient.on("connect", () =>
    console.log("✅ Redis connected successfully")
  );
  redisClient.on("reconnecting", () => console.log("Reconnecting to Redis..."));

  try {
    await redisClient.connect();
    console.log(`✅ Redis Connected: ${redisUrl}`);
    return redisClient;
  } catch (error) {
    console.warn(`⚠️ Redis connection failed at ${redisUrl}:`, error.message);
    console.warn(
      "⚠️ Continuing without Redis - rate limiting will use memory store"
    );
    redisClient = null; // Reset to null so getRedisClient knows it's not available
    return null;
  }
};

/**
 * Get the Redis client instance
 * @returns {object|null} Redis client or null if not available
 */
const getRedisClient = () => {
  return redisClient; // Return null if not initialized or connection failed
};

/**
 * Close Redis connection
 */
const closeRedisClient = async () => {
  if (redisClient) {
    await redisClient.quit();
    redisClient = null;
  }
};

/**
 * Connect to Redis - similar pattern to connectDB
 * @returns {Promise<object|null>} Redis client or null if connection fails
 */
const connectRedis = async () => {
  try {
    const client = await initRedisClient();
    return client;
  } catch (error) {
    console.warn("⚠️ Redis connection failed:", error.message);
    console.warn(
      "⚠️ Continuing without Redis - rate limiting will use memory store"
    );
    return null;
  }
};

module.exports = {
  connectRedis,
  initRedisClient,
  getRedisClient,
  closeRedisClient,
};

File: ai-service/config/redis_client.py
"""
Redis configuration and connection management for FastAPI service
"""

import os
import logging
import redis.asyncio as redis
from typing import Optional

logger = logging.getLogger(__name__)

# Global Redis client
redis_client: Optional[redis.Redis] = None
redis_connection_status = "initializing"


def get_redis_url() -> str:
    """
    Determines the Redis connection URL.
    Uses REDIS_URL from environment if available.
    Defaults to 'redis://redis:6379' if NODE_ENV is 'production' (Docker context).
    Defaults to 'redis://localhost:6379' otherwise (local development context).
    """
    if os.getenv("REDIS_URL"):
        return os.getenv("REDIS_URL")

    environment = os.getenv("NODE_ENV", "development")
    return (
        "redis://redis:6379"
        if environment == "production"
        else "redis://localhost:6379"
    )


async def init_redis_client() -> Optional[redis.Redis]:
    """Initialize async Redis client (singleton)"""
    global redis_client, redis_connection_status

    if redis_client:
        return redis_client

    redis_url = get_redis_url()
    logger.info(f"Attempting to connect to Redis at: {redis_url}")

    try:
        redis_client = redis.from_url(
            redis_url,
            encoding="utf-8",
            decode_responses=True,
        )
        await redis_client.ping()
        redis_connection_status = "connected"
        logger.info(f"Redis Connected: {redis_url}")
        return redis_client

    except Exception as error:
        logger.warning(f"Redis connection failed at {redis_url}: {error}")
        logger.warning("Continuing without Redis - caching will be disabled")
        redis_client = None
        redis_connection_status = "disconnected"
        return None


async def get_redis_client() -> Optional[redis.Redis]:
    """Get the async Redis client instance"""
    if redis_client:
        return redis_client
    return await init_redis_client()


def get_redis_connection_status() -> str:
    """Get current Redis connection status"""
    return redis_connection_status


def close_redis_client():
    """Close Redis connection"""
    global redis_client, redis_connection_status

    if redis_client:
        try:
            redis_client.close()
            logger.info("Redis connection closed")
        except Exception as e:
            logger.error(f"Error closing Redis connection: {e}")
        finally:
            redis_client = None
            redis_connection_status = "disconnected"
